Add method to transform ClassObjectMethod to code

MightyZanark opened the pull request:
- Add `to_views_code` method on ClassObjectMethod to transform it into standard Python code
- Add `to_views_code` method on ParameterObject to transform it into standard Python function parameter
- Add getters for attributes on ClassObjectMethod and TypeObject

Not0nline commented:
### Halo, setelah mereview kode ini, mungkin ada beberapa improvement yang bisa dibuat

1. **Single Responsibility Principle (SRP)**
- Membuat agar to_views_code pada kelas ClassMethodObject mengikuti aturan Single Responsibility Principle (SRP) karena yang dilakukan sekarang, to_views_code melakukan generasi kode python, validasi, dan juga store data, sedangkan seharusnya kalau ingin mengikuti SRP, to_views_code seharusnya hanya melakukan generasi kode python. Validasi dan store data seharusnya dilakukan oleh fungsi lain
- **Solusi**: membuat fungsi baru yang dapat melakukan validasi dan juga store data agar mengikuti arutan SRP
- **Improvement untuk tim**: hal ini dapat membantu tim karena dapat mengubah validasi inpur tanpa impact satu dengan yang lain, dan juga hal tersebut membuat lebih gampang untuk debug dan update bagian-bagian kode tertentu
2. **Open/Closed Principle (OCP)**  
- kode format baru seperti django views memerlukan modifikasi ClassMethod Object yang melanggar peraturan SOLID bagian Open/Closed Principle (OCP) 
- **Solusi**: Menggunakan Strategy Pattern untuk generasi kode
- **Improvement untuk tim**:  format baru seperti API atau database dapat ditambahkan dengan strategies baru tanpa mengubah kode yang sekarang, dan beberapa generator dapat dibuat bersamaan ketika ingin membuat lebih banyak template selain untuk django
3. **Dependency Inversion Principle (DIP)** 
- ClassMethodObject mempunyai implementasi konkret langsung seperti dari PythonViewsGenerator.
- **Solusi**: menurunkan implementasi dari abstractions (interfaces) 
- **Improvement untuk tim**: mengurangi coupling sehingga lebih gampang untuk menukar-nukar implementasi seperti testing dengan mock generator, dan unit tests dapat inject mock generators untuk mengisolasikan behaviournya
4. Factory Pattern
- Membuat object comples seperti ClassMethodObject dengan dependency yang banyak dependencies secara manual memerlukan resource yang banyak dan memiliki kemungkinan error yang tinggi.
- **Solusi**: Menggunakan factory untuk centralisasi pembuatan object.
- **Improvement untuk tim**: memastikan semua ClassMethodObject instances yang dibuat mempunyai value default yang valid, dan karena creation logic tercentralisasi maka, meminimalkan kemungkinan terjadi kesalahan seperti pada setup secara manual

### Untuk alat-alat yang dapat digunakan untuk meningkatkan produktifitas, berikut berberapa sugestion saya:
- **Linters/Formatter**:  
  Integrasikan `flake8` dan `Black` untuk menegakkan kepatuhan PEP 8 dan memformat kode otomatis selama commit.
- **Type Checkers**:  
  Gunakan `mypy` untuk memvalidasi type hints di output `to_views_code` (misalnya, tipe return `-> str`).
- **Dokumentasi**:  
  Hasilkan dokumentasi API dengan Sphinx untuk menjelaskan tujuan `to_views_code` dan getter kepada anggota tim lain.

KingOnyxx commented:
This is a solid addition of code—the new to_views_code method and the accompanying getters add useful functionality. I have a suggestion to further refine the implementation regarding the to_views_code methods

- Refactor for Clarity:
The current implementation of to_views_code appears to mix code generation, validation, and data manipulation as mentioned by @Not0nline. Probably best to split them to simplify the function and make it easier to maintain and test. Also, i think isolating the code generation logic in its own dedicated function is worth to consider.

Other than that, the code looks good to me. Good Job!

MightyZanark commented:
Thank you for the reviews, I added some validations because I wasn't sure if the data inside those classes by the time they need to be converted to code would be valid or not. For the time being, I will make a function to check the validity of the data so that the `to_views_code` doesn't need to handle it. If in the future the data gets validated from someplace else before being ready to get turned to code, then I will refactor or deprecate the validations here.

DaWanAnOnli commented:
I think the suggestions made by @KingOnyxx  and @Not0nline are fantastic. I just want to add one note. As these intermediate classes are meant to be scalable (i.e. able to be used by many web frameworks, not just Django), I suggest that the names of the methods to transform these classes into Django code should contain Django in them, to differentiate from other similar methods that will map these classes to different frameworks. 

For example:
to_views_code can be renamed to be: to_django_views_code.

That's all from me. It's a subtle change but I believe it's worth it in the long run.
Keep up the good work! Cheers!

MightyZanark commented:
I have fixed some previously not ran assertions so it is ran now and moved the validation out to `utils.py` for now. I haven't completely removed it yet as the parser does not seem fully robust yet, so it is just a layer of precaution.

Regarding the `to_views_code` name and the string concatenation, I decided to hold on to it until we agree on a templating engine to use before completely renaming it to `to_django_views_code` or similar and changing the string concatenation to using the templating engine instead.

Can I merge this part now?

Not0nline commented:
> I have fixed some previously not ran assertions so it is ran now and moved the validation out to `utils.py` for now. I haven't completely removed it yet as the parser does not seem fully robust yet, so it is just a layer of precaution.
> 
> Regarding the `to_views_code` name and the string concatenation, I decided to hold on to it until we agree on a templating engine to use before completely renaming it to `to_django_views_code` or similar and changing the string concatenation to using the templating engine instead.
> 
> Can I merge this part now?

in my opinion yes, not completely removing the validation out of the `utils.py` is fine while the parser is not robust yet because i feel it is better to not change it partially and make the code not work entirely, in my opinion, merging it now wouldn't be a problem as long as the implementation will be changed later when the parser is more robust
